README
Fryktelig fin faktura
=====================

To force the program to start in english, run "LC_ALL=en_US ./faktura".
Some hard coded strings might still be in norwegian, as this program is made exlusively for use in Norway.


Dette er et enkelt program for å lage fakturaer. Fakturaene lages som pdf_ og kan skrives ut på skjema f60 eller sendes som epost. Programmet lager en database over varer, kunder, og utsendte fakturaer.

Programmet startes ved å kjøre «faktura». For tekstmodus, bruk opsjonen -i. Flere opsjoner_ er også mulig.
Miljøvariabelen_ FAKTURADIR kan brukes til å styre hvor databasen skal lagres. Eksempel:

$ FAKTURADIR=/var/fakturadb ./faktura

Du finner mer informasjon om programmet, og informasjon om feilsøking på prosjektsidene_.

.. _pdf: http://en.wikipedia.org/wiki/Portable_Document_Format
.. _opsjoner og miljøvariabler: https://sourceforge.net/p/finfaktura/wiki/Starte%20programmet/
.. _prosjektsidene: https://sourceforge.net/projects/finfaktura/

Endre fakturanummer
===================

Det er nå mulig å endre første fakturanummer, men det virker kun dersom du ikke allerede har laget fakturaer. 
Dette kan kun gjøres én gang. Du kan ikke endre dette tallet senere.
i mappen scripts/ kan du kjøre programmet «./fakturanummer.py» og følge instruksjonene.
Installerte du programmet via en GNU/Linux-pakke, finner du filen her: «/usr/share/finfaktura/scripts/fakturanummer.py»

Diskusjonsgruppe
================

Lurer du på noe? Er det noe som mangler?
Ta en titt på diskusjonsgruppa, hvor nye versjoner og funksjoner blir diskutert, hvor problemer blir løst, og hvor solen aldri går ned:
https://lists.sourceforge.net/lists/listinfo/finfaktura-support

Vi har også en kanal på Freenode: #finfaktura

Systemkrav
==========

Programmet er laget for å kjøre på GNU/Linux, Windows og Apple osx, men utviklerteamet bruker og tester kun i GNU/Linux. Vi kan derfor ikke garantere at det fungerer i andre operativsystemer.
Fryktelig fin faktura utvikles og testes i Mageia med python 2.7.

Følgende moduler er påkrevet:

- python_ 2, versjon 2.5 eller nyere.
- python-modulen PyQt4_, versjon 4.4 eller nyere.
- python-modulen Reportlab_, versjon 2 eller nyere.

Disse oppfylles automatisk når du installerer fra deb/rpm-pakke.

.. _python: http://python.org/
.. _PyQt4: http://www.riverbankcomputing.com/software/pyqt/intro
.. _reportlab: http://www.reportlab.org/rl_toolkit.html


Opphavsrett
===========

Idé og koding: Håvard Gulldahl <havard@gulldahl.no>
Dette er Fri programvare, du kan fritt gjenbruke og videredistribuere under GPL2-lisensen.
Ny prosjektledelse: Johnny A. Solbu <johnny@solbu.net> fra 2015

Se LICENSE eller http://www.gnu.org/licenses/old-licenses/gpl-2.0.html#TOC1


Kjente feil og planlagte egenskaper
===================================

Se https://sourceforge.net/p/finfaktura/tickets/



ChangeLog
v2.2.1 2018-06-18  Johnny A. Solbu <johnny@solbu.net>

	* f60.py: Fiks «Kunne ikke importere PIL»

v2.2.0 2018-02-25  Johnny A. Solbu <johnny@solbu.net>

	* Legg til gui-ikon
	* Legg «Første fakturanummer» i menyen
	* Legg til dokumentasjon i «setup.py bdist_rpm»

v2.1.4 2017-06-06  Johnny A. Solbu <johnny@solbu.net>

	* f60.py: Få tilbake «KVITTERING» på utskrevne kvitteringer

v2.1.3 2016-11-16  Johnny A. Solbu <johnny@solbu.net>

	* gui.py: Ikke krev e-postadresse på kunder. Ikke alle kunder har/oppgir epost

v2.1.2 2015-11-06  Johnny A. Solbu <johnny@solbu.net>

	* setup.py: Installer oversettelser så de fungerer i Linux

v2.1.1 2015-10-30  Johnny A. Solbu <johnny@solbu.net>

	* Ikke hardkode at okular brukes som PDF-program, bruk xdg-open i stedet.

v2.1.0 2015-10-25  Johnny A. Solbu <johnny@solbu.net>

	* Første versjon med støtte for oversettelser. (Takk til Eivind Tagseth)

v2.0.10 2015-10-17  Johnny A. Solbu <johnny@solbu.net>

	* finfaktura/f60.py: Fikser bug 86, hvor reportlab 3 ikke laget fakturaer med fakturafakta.
	* Fjerner siste rest av linker til Google code-siden
	* Ikke krev organisasjonsnummer. Ikke alle som fakturerer har organisasjonsnummer.

v2.0.9 2015-08-17  Johnny A. Solbu <johnny@solbu.net>

	* finfaktura/f60.py: Kontrollsiffer til beløp beregnes på feil grunnlag
	* Fiks Bug 81. Erstat «/» med «_» ved generering av pdf-filer hvor firmanavn har slash i navnet (A/S)
	* Endre navn på kommandofil: faktura.py -> faktura, for korrekt installasjon på GNU/Linux

v2.0.8 2015-08-12  Johnny A. Solbu <johnny@solbu.net>

	* Fiks krasj med UnicodeDecodeError ved oppstart
	* Noen steder har man kun postnummer og sted men ikke veiadresser, ikke krev veiadresse

v2.0.7 2015-07-30  Johnny A. Solbu <johnny@solbu.net>

	* Ikke krev kontaktperson
	* Fiks mellomrom mellom kr og beløp
	* Ny maintainer.

2009-04-12 19:27  havard.gulldahl

	* finfaktura/f60.py: metadata

2009-04-12 19:20  havard.gulldahl

	* finfaktura/f60.py: plass til fem adresselinjer i adressefelt

2009-04-12 18:52  havard.gulldahl

	* finfaktura/f60.py: fikser bug#56, hvor dollar-tegn dukker opp
	  foran valutafelt dersom locale ikke ble satt riktig

2009-04-08 22:26  havard.gulldahl

	* scripts/ff-fakturanummer/fakturanummer.py,
	  scripts/ff-fakturanummer/gui.ui: gui og mesteparten av logikken
	  på plass

2009-04-08 09:53  havard.gulldahl

	* scripts/ff-fakturanummer/fakturanummer.py: mer'

2009-04-06 15:27  havard.gulldahl

	* scripts/ff-fakturanummer,
	  scripts/ff-fakturanummer/fakturanummer.py,
	  scripts/ff-fakturanummer/gui.ui: legger til
	  fiksførstefakturanummer

2009-03-28 21:57  havard.gulldahl

	* finfaktura/gui.py: skjuler vare- og kundeinfo når de skal
	  redigeres - bedre for små vinduer

2009-03-28 21:42  havard.gulldahl

	* finfaktura/fakturabibliotek.py: fikser issue#55 - statusfelt med
	  ikke-ascii-tegn

2009-03-28 21:23  havard.gulldahl

	* finfaktura/f60.py: fikser bug hvor linjeskift i fakturatekst ble
	  erstattet med svart firkant (issue#54)

2009-03-22 21:55  havard.gulldahl

	* debian/control, debian/finfaktura.menu, doc/REGLER,
	  finfaktura-installer.nsi: bedre installasjonsinstillinger på win
	  og debian

2009-03-22 21:51  havard.gulldahl

	* dist, dist/finfaktura-win32-base-2_0_x.zip, dist/library.zip:
	  legger inn win-kode for netinstaller

2009-03-16 09:27  havard.gulldahl

	* finfaktura-installer.nsi: mer fiks

2009-03-16 03:35  havard.gulldahl

	* finfaktura-installer.nsi: fiks nedlastingslogik

2009-03-16 03:19  havard.gulldahl

	* finfaktura-installer.nsi: webinstaller

2009-03-12 22:22  havard.gulldahl

	* ChangeLog, debian/changelog, finfaktura/__init__.py: Gjør 2.0.5
	  klar til utgivelse

2009-02-17 22:38  havard.gulldahl

	* finfaktura/f60.py: API-ENDRING: legger til argumentet lvertEpoch
	  til metoden f60.settFakturainfo()

2009-02-17 22:33  havard.gulldahl

	* finfaktura/f60.py: bruker en intern metode for å skrive
	  valutabeløp, for å være på den sikre siden

2009-02-17 22:21  havard.gulldahl

	* finfaktura/f60.py: forsøk på bedre locale-velger

2009-02-17 22:20  havard.gulldahl

	* finfaktura/test.py: test for kid

2009-02-12 20:36  havard.gulldahl

	* finfaktura/f60.py: ikke snuble på filsjekk av logo

2009-02-12 13:27  havard.gulldahl

	* finfaktura/f60.py: massive forbedringer av blanketten (stor takk
	  til cbrattli): plassering av ulike felter er korrigert,
	  blanketten skrives mer fullstendig. Også: bedre valg av locale.
	  Fiks i mod10-algoritme - igjen, takk til cbrattli. fiks utskrift
	  av kontonummer som starter på 0 (igjen)

2009-02-11 12:16  havard.gulldahl

	* finfaktura/f60.py: fikser typos

2009-02-10 11:33  havard.gulldahl

	* finfaktura/f60.py: kan sette begrensninger på bruken av pdf-en
	  (krever reportlab > 2.3)

2009-02-08 02:06  havard.gulldahl

	* finfaktura/f60.py: tar vare på høyde-bredde-aspektet i logoen når
	  den settes inn (ikke strekk den begge veger hvis det ikke er
	  nødvendig)

2009-02-08 01:48  havard.gulldahl

	* finfaktura/f60.py: mer robust(?) måte å laste logoer på

2009-02-08 00:57  havard.gulldahl

	* finfaktura/f60.py: forbedrer logo-rutinen. kan lese jpg direkte.
	  må teste PIL

2009-02-07 23:58  havard.gulldahl

	* finfaktura/f60.py: Setter pdf-metadata

2009-02-07 23:48  havard.gulldahl

	* finfaktura/f60.py: Regner ut og spesifiserer momsgrunnlag

2009-02-07 22:27  havard.gulldahl

	* finfaktura/f60.py: bruker locale.currency() på alle beløp, slik
	  at de formatteres med riktige skilletegn og prefikses med kr

2009-02-06 00:41  havard.gulldahl

	* finfaktura/f60.py: bruker mod10 for å generere KID

2009-02-06 00:40  havard.gulldahl

	* finfaktura/f60.py: mer nøyaktig angivelse av den nederste
	  tekstlinja på blanketten, jf issue#38, punkt A

2009-02-06 00:24  havard.gulldahl

	* finfaktura/f60.py: innstillingsmerket H (issue#38, punkt A)

2009-02-05 23:41  havard.gulldahl

	* finfaktura/f60.py: nå også kontrollsiffer etter mod11-algoritmen
	  hos BBS (issue#38). Fjerner feilslått filnavnfiks. skriver

finfaktura/gui_firma.py
# -*- coding: utf-8 -*-
# kate: indent-width 4;
###########################################################################
#    Copyright (C) 2008 Håvard Gulldahl
#    <havard@gulldahl.no>
#
#    Lisens: GPL2
#
# $Id: faktura.py 260 2008-05-11 08:59:23Z havard.gulldahl $
#
###########################################################################
 
import logging
from PyQt4 import QtCore, QtGui
from ui import firmainfo_ui
from fakturabibliotek import typeofqt

class firmaOppsett(firmainfo_ui.Ui_firmaOppsett):
    def __init__(self, firma):
        self.firma = firma
        self.gui = QtGui.QDialog()
        self.setupUi(self.gui)
        self.gui.connect(self.lagreLogo, QtCore.SIGNAL("clicked()"), self.finnFjernLogo)

        self._kontrollkart = {
            self.Firmanavn:'Firmanavn',
            #self.Organisasjonsnummer:u'Organisasjonsnummer fra Brønnøysund',
            self.Kontaktperson:'Kontaktperson',
            self.Epost:'Epostadresse',
            self.Adresse:'Adresse',
            self.Postnummer:'Postnummer',
            self.Poststed:'Poststed',
            self.Telefon:'Telefonnummer',
            self.Mobil:'Mobilnummer',
            self.Kontonummer:'Kontonummer',
            #self.Mva:'Momssats',
            self.Forfall:'Forfallsperiode',
        }
        self.gui.show()
        self.vis()
        #self.visLogo()


    def exec_(self):
        res = self.gui.exec_()
        if res == QtGui.QDialog.Accepted:
            return self.samleInfo()
        return {}

############## FIRMAINFO ###################

    def firmaWidgetKart(self):
        return {
            self.Firmanavn            :  self.firma.firmanavn,
            self.Organisasjonsnummer  :  self.firma.organisasjonsnummer,
            self.Kontaktperson        :  self.firma.kontaktperson,
            self.Epost                :  self.firma.epost,
            self.Adresse              :  self.firma.adresse,
            self.Postnummer           :  self.firma.postnummer,
            self.Poststed             :  self.firma.poststed,
            self.Telefon              :  self.firma.telefon,
            self.Mobil                :  self.firma.mobil,
            self.Telefaks             :  self.firma.telefaks,
            self.Kontonummer          :  self.firma.kontonummer,
            self.Vilkar               :  self.firma.vilkar,
            self.Mva                 :  self.firma.mva,
            self.Forfall             :  self.firma.forfall,
            }

    def vis(self):
        format = { self.Postnummer: "%04i", self.Kontonummer: "%011i", } # må formateres spesielt dersom det begynner med 0, eks. 0921
        for til, fra in self.firmaWidgetKart().iteritems():
            #debug("fra", fra, type(fra))
            #debug("til", til, type(til))
            if hasattr(til, 'setText'):
                if not fra: continue
                if til in format: s = format[til] % fra
                else: s = unicode(fra)
                til.setText(s)
            elif hasattr(til, 'setValue'):
                if not fra: continue
                til.setValue(int(fra))
            elif hasattr(til, 'setPlainText'):
                if not fra: continue
                til.setPlainText(unicode(fra))
        self.visLogo()
        self.firmaSjekk()

    def visLogo(self):
        if not self.firma.logo:
            self.lagreLogo.setText('&Finn logo')
            #self.LogoPixmap.setPixmap(QtGui.QPixmap())
            self.LogoPixmap.clear()
        else:
            #pass
            logging.debug('visLogo: %s, %s', repr(self.firma.logo), len(self.firma.logo))
            logo = QtGui.QPixmap()
            logo.loadFromData(self.firma.logo)
            self.LogoPixmap.setPixmap(logo)
            self.lagreLogo.setText('&Fjern logo')

    def kanskjetall(self, obj):
        try:
            return int(obj.text())
        except ValueError:
            return None

    def samleInfo(self):
        r = {}
        r['firmanavn'] = unicode(self.Firmanavn.text())
        r['organisasjonsnummer'] = unicode(self.Organisasjonsnummer.text())
        r['kontaktperson'] = unicode(self.Kontaktperson.text())
        r['epost'] = unicode(self.Epost.text())
        r['adresse'] = unicode(self.Adresse.toPlainText())
        r['postnummer'] = self.kanskjetall(self.Postnummer)
        r['poststed'] = unicode(self.Poststed.text())
        r['telefon'] = self.kanskjetall(self.Telefon)
        r['mobil'] = self.kanskjetall(self.Mobil)
        r['telefaks'] = self.kanskjetall(self.Telefaks)
        r['kontonummer'] = self.kanskjetall(self.Kontonummer)
        r['vilkar'] = unicode(self.Vilkar.toPlainText())
        r['mva'] = int(self.Mva.value())
        r['forfall'] = int(self.Forfall.value())
        return r

    def oppdater(self):
        self.firma.firmanavn  = unicode(self.Firmanavn.text())
        self.firma.organisasjonsnummer = unicode(self.Organisasjonsnummer.text())
        self.firma.kontaktperson = unicode(self.Kontaktperson.text())
        self.firma.epost      = unicode(self.Epost.text())
        self.firma.adresse    = unicode(self.Adresse.toPlainText())
        self.firma.postnummer = self.kanskjetall(self.Postnummer)
        self.firma.poststed   = unicode(self.Poststed.text())
        self.firma.telefon    = self.kanskjetall(self.Telefon)
        self.firma.mobil      = self.kanskjetall(self.Mobil)
        self.firma.telefaks   = self.kanskjetall(self.Telefaks)
        self.firma.kontonummer = self.kanskjetall(self.Kontonummer)
        self.firma.vilkar     = unicode(self.Vilkar.toPlainText())
        self.firma.mva        = int(self.Mva.value())
        self.firma.forfall    = int(self.Forfall.value())
        mangler = self.sjekkFirmaMangler()
        if mangler:
            mangel = u'Du er nødt til å oppgi:\n%s' % ([ mangler[obj].lower() for obj in mangler.keys() ])
            logging.debug (mangel)
            QtGui.QMessageBox.critical(self.gui, 'Ufullstendige opplysninger', mangel)
            mangler.keys()[0].setFocus()
            return False

    def sjekkFirmaMangler(self):
        kravkart = {}
        kravkart.update(self._kontrollkart)
        test = None
        for obj in kravkart.keys():
            if isinstance(obj, QtGui.QSpinBox): test = obj.value() > 0
            elif isinstance(obj, QtGui.QComboBox): test = obj.currentText()
            elif isinstance(obj, (QtGui.QLineEdit,QtGui.QTextEdit,)): test = obj.text()
            if test is None:
                logging.error('sjekkFirmaMangler: mangler test for % obj')
            elif test: kravkart.pop(obj)
        return kravkart

    def firmaSjekk(self, event=None):
        ok = QtGui.QColor('white')
        tom = QtGui.QColor('red')
        widget = 'QWidget'
        for obj in self._kontrollkart.keys():
            if isinstance(obj, (QtGui.QSpinBox, QtGui.QDoubleSpinBox)): test = obj.value() > 0
            elif isinstance(obj, QtGui.QComboBox): test = obj.currentText()
            elif isinstance(obj, (QtGui.QLineEdit,QtGui.QTextEdit,)): test = obj.text()
            elif isinstance(obj, (QtGui.QPlainTextEdit,)): test = obj.toPlainText()
            else:
                logging.error('mangler test for %s (%s)' % (obj, type(obj)))
            if test:
                #obj.setPaletteBackgroundColor(ok)
                #logging.debug('firmaSjekk: ok for %s', obj.typeName())
                obj.setStyleSheet("%s { background-color: white; }" % typeofqt(obj))
            else:
                obj.setStyleSheet("%s { background-color: red; }" % typeofqt(obj))
                #obj.setPaletteBackgroundColor(tom)

    def finnFjernLogo(self):
        logging.debug('finnFjernLogo %s', repr(self.firma.logo))
        logging.debug('finnFjern lengde %s', len(str(self.firma.logo)))
        if (self.firma.logo):
            logging.debug('fjerner logo %s', repr(self.firma.logo))
            self.firma.logo = ''
            self.visLogo()
        else:
            startdir = ""
            logofile = QtGui.QFileDialog.getOpenFileName(self.gui,
                "Velg bildefil for firmaets logo",
                startdir,
                'Bildefiler (*.png *.xpm *.jpg *.jpeg *.gif *.bmp *.ppm *.pgm *.pbm)',
                )
            if len(unicode(logofile)) > 0:
                logging.debug ("Setter ny logo: %s" % logofile)

                logo = QtGui.QPixmap(logofile)
                if logo.isNull(): # kunne ikke åpne logo
                    return False

                stream = QtCore.QBuffer()
                scaledlogo = logo.scaled(QtCore.QSize(360,360), QtCore.Qt.KeepAspectRatio, QtCore.Qt.SmoothTransformation)
                scaledlogo.save(stream, 'PNG')
                #self.firma.logo = buffer(stream.data())
                self.firma.logo = buffer(str(stream.data()))
                logging.debug('logo: %s %s %s', type(self.firma.logo), len(self.firma.logo), repr(self.firma.logo))
                self.visLogo()

                #self.firma.hentEgenskaper()
                #logging.debug('hentet egenskaper. logo: %s, %s', type(self.firma.logo), len(self.firma.logo))